### DISCUSSION TITLE
What is correct way to get block strings now?

### DISCUSSION THREAD

**Discussion by Niklan** [thephpleague/commonmark#650] [Q&A] [answered]
Hi.

Previously in my block renderer, I was using `$strings = $block->getStrings();`, now this method is gone and I don't see it in the changelog.

I use this to get content on my block, which also has:

```php
  public function canContain(AbstractBlock $block): bool {
    return FALSE;
  }
```

in element definition.

I'm 100% sure there will be only string inside a block, and now I can't access it.

This is what I get in `render()` method from `$block`.

![Screenshot_20190420_133758](https://user-images.githubusercontent.com/2356744/56454911-84160280-6371-11e9-9aab-0cc7f80ed500.png)

Why does it in "next" Paragraph, and sometimes in Heading.

### Element

```php
<?php

namespace Drupal\druki_parser\CommonMark\Block\Element;

use League\CommonMark\Block\Element\AbstractBlock;
use League\CommonMark\ContextInterface;
use League\CommonMark\Cursor;
use League\CommonMark\Util\RegexHelper;

/**
 * Class MetaInformationElement
 *
 * @package Drupal\druki_parser\Plugin\Markdown\Extension
 */
class MetaInformationElement extends AbstractBlock {

  /**
   * Indicates is meta information closing block found.
   *
   * @var bool
   */
  protected $isCloserFound;

  /**
   * MetaInformation constructor.
   */
  public function __construct() {
    $this->isCloserFound = FALSE;
  }

  /**
   * {@inheritdoc}
   */
  public function canContain(AbstractBlock $block): bool {
    return FALSE;
  }

  /**
   * {@inheritdoc}
   */
  public function acceptsLines(): bool {
    return TRUE;
  }

  /**
   * {@inheritdoc}
   */
  public function isCode(): bool {
    // @todo Avoid it. Not sure this is good approach, but only working with
    // matches. Try some workaround with $context->advanceBy(-3) in
    // handleRemainingContents(). This will let condition pass. But need some
    // extra fixes, since this not fix all the problems that fix this value.
    return TRUE;
  }

  /**
   * {@inheritdoc}
   */
  public function shouldLastLineBeBlank(Cursor $cursor, $currentLineNumber): bool {
    return FALSE;
  }

  /**
   * {@inheritdoc}
   */
  public function matchesNextLine(Cursor $cursor): bool {
    if ($this->isCloserFound()) {
      return FALSE;
    }

    return TRUE;
  }

  /**
   * @return bool
   */
  public function isCloserFound(): bool {
    return $this->isCloserFound;
  }

  /**
   * @param bool $isCloserFound
   */
  public function setIsCloserFound($status): void {
    $this->isCloserFound = $status;
  }

  /**
   * {@inheritdoc}
   */
  public function handleRemainingContents(ContextInterface $context, Cursor $cursor): void {
    if ($cursor->getNextNonSpaceCharacter() == '-') {
      $match = RegexHelper::matchAll('/^\-{3}$/', $cursor->getLine(), $cursor->getNextNonSpacePosition());
      if (!empty($match)) {
        $this->setIsCloserFound(TRUE);
        return;
      }
    }

    $context->getTip()->addLine($cursor->getRemainder());
  }

}
```

### Parser

```php
<?php

namespace Drupal\druki_parser\CommonMark\Block\Parser;

use Drupal\druki_parser\CommonMark\Block\Element\MetaInformationElement;
use League\CommonMark\Block\Parser\BlockParserInterface;
use League\CommonMark\ContextInterface;
use League\CommonMark\Cursor;

/**
 * Class MetaInformationParser
 *
 * @package Drupal\druki_parser\CommonMark\Block\Parser
 */
class MetaInformationParser implements BlockParserInterface {

  /**
   * {@inheritdoc}
   */
  public function parse(ContextInterface $context, Cursor $cursor): bool {
    // Only works for metadata found at the beginning of the file. Other "---"
    // will be replaced as expected with "<hr />".
    if ($context->getLineNumber() > 1) {
      return FALSE;
    }

    if ($cursor->isIndented()) {
      return FALSE;
    }

    $meta_information = $cursor->match("/^\-{3}$/");
    if (!$meta_information) {
      return FALSE;
    }

    $context->addBlock(new MetaInformationElement());

    return TRUE;
  }

}
```

### Renderer

```php
<?php

namespace Drupal\druki_parser\CommonMark\Block\Renderer;

use Drupal\Core\Serialization\Yaml;
use League\CommonMark\Block\Element\AbstractBlock;
use League\CommonMark\Block\Renderer\BlockRendererInterface;
use League\CommonMark\ElementRendererInterface;
use League\CommonMark\HtmlElement;

/**
 * Class MetaInformationRenderer
 *
 * @package Drupal\druki_parser\CommonMark\Renderer
 */
class MetaInformationRenderer implements BlockRendererInterface {

  /**
   * {@inheritdoc}
   */
  public function render(AbstractBlock $block, ElementRendererInterface $htmlRenderer, $inTightList = FALSE): HtmlElement {
    $strings = $block->getStrings();
    $content = [];

    $yaml_string = implode($strings, PHP_EOL);
    $yaml_array = Yaml::decode($yaml_string);
    foreach ($yaml_array as $key => $value) {
      $content[] = new HtmlElement('div', [
        'data-druki-key' => $key,
        'data-druki-value' => is_array($value) ? implode(', ', $value) : $value,
      ]);
    }

    return new HtmlElement('div', ['data-druki-meta' => ''], $content);
  }

}
```

### Content example

```markdown
---
id: route-access-control
title: Контроль доступа
path: /docs/8/routing/access-control
core: 8
category-area: Маршрутизация
category-order: 4
search-keywords:
  - доступ защита маршрутов контроллеров права доступа
---
```

Can you point me in the right direction?

**Comment by colinodell** [comment, marked as answer]
> Previously in my block renderer, I was using $strings = $block->getStrings();, now this method is gone and I don't see it in the changelog.

My apologies, it looks like I forgot to document this change.  I will fix this ASAP.

Basically, you'll want to make these changes:

1. Have your `MetaInformationElement` class extend from `AbstractStringContainerBlock` instead of `AbstractBlock`
2. Override its `finalize()` method to set the final contents like `$this->finalStringContents = implode("\n", $this->strings);`
3. Call the `getStringContent()` method in your renderer to get those finalized string contents

**Comment by Niklan** [comment]
Thank you for the fast response. I'll check it and reply again. 👍

**Comment by Niklan** [comment]
Fixed it! Thanks.

I have another question since I asked already. In my element (same example) I need to set `isCode()` to `TRUE`, or parser will not work correctly.

For example, we  have content to parse:

```
---
id: some-id
yaml-array:
  - one
  - two
---
```

With `isCode() = TRUE` the result string in rednerer will be

```
id: some-id\n
yaml-array:\n
  - one\n
  - two
```

So, this is parsed correctly. Especially array.

But when I set `isCode()` to `FALSE`, the result will be:

```
id: some-id\n
yaml-array:
```

As you can see, YAML array is missing. I think this is parsed something like lists, but don't know why, because I still have `canContain()` `FALSE`.

I can't understand the difference. Need I fix it, or my "hack" is good solution?

**Comment by colinodell** [comment]
Perhaps a better name for `isCode()` would be `isBlockExtraGreedy()` - when `TRUE`, the engine will assume that most of the subsequent lines of Markdown belong to your block (until it says otherwise) - it won't try parsing those lines with other parsers until you say (via `matchesNextLine()`) that your code/"greedy" block no longer matches.  So yes - your approach makes sense to me!

Because this type of greedy behavior mostly belongs to code elements, we call it `isCode()`, as does the reference JavaScript parser.  I'll consider whether we want to rename this to make its purpose more clear.  I'll also explain this behavior in the documentation.

**Comment by colinodell** [comment]
Additional documentation has been added :)  I'm going to close this for now, but let me know if you have any further questions I can assist with!

**Comment by Niklan** [comment]
Thank you again. Now I understand what `isCode()` does. Actually, I think it makes sense to rename it or make an alias to `isBlockExtraGreedy()`. Because as we can see, not only the code is the only use case for it. The current name is misleading, but documentation makes it clear now.
